apmproxy/apmserver.go

```diff
@@ -40,27 +40,20 @@ type jsonError struct {
 	Document string `json:"document,omitempty"`
 }
 
-// ForwardApmData receives agent data as it comes in and posts it to the APM server.
-// Stop checking for, and sending agent data when the function invocation
+// ForwardApmData receives apm data as it comes in and posts it to the APM server.
+// Stop checking for, and sending apm data when the function invocation
 // has completed, signaled via a channel.
-func (c *Client) ForwardApmData(ctx context.Context, metadataContainer *MetadataContainer) error {
+func (c *Client) ForwardApmData(ctx context.Context) error {
 	if c.IsUnhealthy() {
 		return nil
 	}
 	for {
 		select {
 		case <-ctx.Done():
 			c.logger.Debug("Invocation context cancelled, not processing any more agent data")
-			return nil
-		case agentData := <-c.DataChannel:
-			if metadataContainer.Metadata == nil {
-				metadata, err := ProcessMetadata(agentData)
-				if err != nil {
-					return fmt.Errorf("failed to extract metadata from agent payload %w", err)
-				}
-				metadataContainer.Metadata = metadata
-			}
-			if err := c.PostToApmServer(ctx, agentData); err != nil {
+			return c.sendBatch(ctx)
```

Comment:
If the context is cancelled, then shouldn't this mean `sendBatch` should not do anything?

Looking at the sendBatch method, it calls PostToApmServer which makes an HTTP request to APM Server. The HTTP request does not use the provided context, hence why `sendBatch` doesn't get cancelled immediately.

Perhaps on context cancellation here we should use a new background context with a shutdown timeout. WDYT?

Reply:
Ah, thanks for catching this. It is the remnant from when I was creating local batch and to avoid data-loss, flushing them on ctx done. I will fix this to not do anything on context canceled since now the batch is global.

Reply:
Fixed this to not do anything on context done. The extension will rely on `FlushAPMData` to take care of anything in the batch.